File: src/main/java/at/searles/lexer/utils/Interval.java
package at.searles.lexer.utils;

public class Interval implements Comparable<Interval> {

    public final int start;  // start and end are public for optimizations
    public final int end;    // end is exclusive

    public Interval(int ch) {
        this.start = ch;
        this.end = ch + 1;
    }

    public Interval(int start, int end) {
        this.start = start;
        this.end = end;
    }

    @Override
    public int compareTo(Interval that) {
        // Lexical order
        int cmp = Integer.compare(this.start, that.start);

        return cmp != 0 ? cmp : Integer.compare(this.end, that.end);
    }

    @Override
    public boolean equals(Object o) {
        if (getClass() == o.getClass()) {
            return compareTo((Interval) o) == 0;
        }

        return false;
    }

    public int contains(int ch) {
        return ch < start ? -1 : ch >= end ? 1 : 0;
    }

    @Override
    public String toString() {
        return String.format("[%d, %d)", start, end);
    }

}

File: src/main/java/at/searles/lexer/utils/Interval.kt
package at.searles.lexer.utils

data class Interval(val start: Int, val end: Int) : Comparable<Interval> {
    constructor(ch: Int): this(ch, ch + 1)

    init {
        require(start < end)
    }

    override fun compareTo(other: Interval): Int {
        // Lexical order
        val cmp = start.compareTo(other.start)
        return if (cmp != 0) cmp else end.compareTo(other.end)
    }

    operator fun contains(ch: Int): Boolean {
        return ch in start until end
    }

    override fun toString(): String {
        return "[$start, $end)"
    }

    companion object {
        val all by lazy {
            Interval(Int.MIN_VALUE, Int.MAX_VALUE)
        }
    }
}